libpyvinyl/BaseData.py

```diff
@@ -0,0 +1,477 @@
+            self.__file_format_class = value
+        elif value is None:
+            self.__file_format_class = None
+        else:
+            raise TypeError(
+                f"Data Class: format_class should be None or a format class, not {type(value)}"
+            )
+
+    @property
+    def file_format_kwargs(self):
+        """The kwargs needed to map the file"""
+        return self.__file_format_kwargs
+
+    @file_format_kwargs.setter
+    def file_format_kwargs(self, value):
+        if isinstance(value, dict):
+            self.__file_format_kwargs = value
+        elif value is None:
+            self.__file_format_kwargs = None
+        else:
+            raise TypeError(
+                f"Data Class: file_format_kwargs should be None or a dict, not {type(value)}"
+            )
+
+    @property
+    def mapping_type(self):
+        """If this data class is a file mapping or python dict mapping."""
+        return self.__check_mapping_type()
+
+    def __check_mapping_type(self):
+        """Check the mapping_type of this class."""
+        if self.data_dict is not None:
+            return dict
+        elif self.filename is not None:
+            return self.file_format_class
+        else:
+            raise TypeError("Neither self.__data_dict or self.__filename was found.")
+
+    @property
+    def mapping_content(self):
+        """Returns an overview of the keys of the mapped dict or the filename of the mapped file"""
+        if self.mapping_type == dict:
+            return self.data_dict.keys()
+        else:
+            return self.filename
+
+    @staticmethod
+    def _add_ioformat(format_dict, format_class):
+        """Register an ioformat to a `format_dict` listing the formats supported by this DataClass.
+
+        :param format_dict: The dict listing the supported formats.
+        :type format_dict: dict
+        :param format_class: The FormatClass to add.
+        :type format_class: class
+        """
+        register = format_class.format_register()
+        for key, val in register.items():
+            if key == "key":
+                this_format = val
+                format_dict[val] = {}
+            else:
+                format_dict[this_format][key] = val
+
+    @classmethod
+    @abstractmethod
+    def supported_formats(self):
+        format_dict = {}
+        # Add the supported format classes when creating a concrete class.
+        # See the example at `tests/BaseDataTest.py`
+        self._add_ioformat(format_dict, FormatClass)
+        return format_dict
+
+    @classmethod
+    def list_formats(self):
+        """Print supported formats"""
+        out_string = ""
+        supported_formats = self.supported_formats()
+        for key in supported_formats:
+            dicts = supported_formats[key]
+            format_class = dicts["format_class"]
+            if format_class != "":
+                out_string += "Format class: {}\n".format(format_class)
```

Comment:
The `.format` used here is because of this line:
https://github.com/PaNOSC-ViNYL/libpyvinyl/blob/6be653a2b4386f436d9dff6700c5e023d24c2985/libpyvinyl/BaseData.py#L226
`f"Description: {dicts["description"]}\n"` doesn't work in my test: `SyntaxError: invalid syntax`.

For consistency, I'm inclined to use`.format` in this block.

Reply:
oh, that's solvable with simply using single quotes inside the f-string:
´f"Description: {dicts['description']}\n"´
should work. The double quotes end the main string, hence the messed up syntax.